onnxruntime/contrib_ops/cpu/bert/multihead_attention.cc

```diff
@@ -323,12 +309,12 @@ Status MultiHeadAttention<T>::Compute(OpKernelContext* context) const {
     TensorShape q_BNSH(new_dims_span);
     Tensor::InitOrtValue(element_type, q_BNSH, allocator, Q);
     if (qkv_bias == nullptr) {
-      Tensor* query_reshaped = nullptr;
+      std::unique_ptr<Tensor> query_reshaped;
```

Comment:
The code for transpose Q, K, V from BSD to BNSH are quite similar. It is possible to refactor it to avoid writing similar code 3 times. 

Also, I think there is no need to allocate a Tensor object. Just pass a TensorShape for shape, and raw pointer of data buffer.

Reply:
>  I think there is no need to allocate a Tensor object
It is not I want to create a Tensor view, it is the underlying API the function will call want to receive a Tensor:
https://github.com/microsoft/onnxruntime/blob/fc413a41fa19158c720068f6b2efc0e7a94ac01d/onnxruntime/contrib_ops/cpu/bert/multihead_attention.cc#L65

Reply:
@tianleiwu This change is coming from commit: "Fix Invalid rank for input: query Got: 4 Expected: 3 Please fix either the inputs or the model." You can see this commit for more info. I only made minimal change to fix.

Should I do the refactoring here or maybe in another PR?

Reply:
Since you touched this code, I suggest refactoring either in this PR or another PR.

Reply:
OK, let's me do it in another PR.